### ISSUE TITLE
Annotations usage across the framework

### ISSUE THREAD

**Issue by cesards** [google/ExoPlayer#1928]
First of all, very good job with the framework. It's very extensible and the API is not overcomplicated and accessible. However it's very difficult to know what components I need to use or not. 

For instance, creating an instance of the player requires some components but "I think" I won't need some of them, so it would be nice using annotations in all fields by your side. The policy used for it (only using `@Nullable` for avoidable fields, or use `@NonNull` and `@Nullable` extensively) I think it's up to you, but it would help a lot to the developers).

Great job and keep working hard, guys!

**Comment by ojw28** [comment]
Yes, this is on our list. It's not high priority at this point, but we'll get to it.

**Comment by ojw28** [comment]
This is blocked on the release of support library 25.1.

**Comment by andrewlewis** [comment]
We've started to use `@Nullable` and `@IntDef` across ExoPlayer's API surface. I think we also intend to start using `@CallSuper`.

There was a plan to apply `@NonNull` by default using a package annotation, but our preferred way of doing this would require support to be added to Android Studio, as Java annotations don't support this inherently (internal bug reference b/36468797). So that plan is on hold for the moment.

I'll close this as the current annotation usage should meet the requirement of documenting optional fields, but please feel free to comment if there are particular places where annotations still need to be added. Thanks!

**Comment by cesards** [comment]
Sweet! Thanks guys!

**Comment by cesards** [comment]
Thanks guys! Appreciated it!!
